hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/AbstractHoodieWriteClient.java

```diff
@@ -708,15 +708,25 @@ public HoodieCleanMetadata clean(String cleanInstantTime, boolean skipLocking) t
    * @param skipLocking if this is triggered by another parent transaction, locking can be skipped.
    */
   public HoodieCleanMetadata clean(String cleanInstantTime, boolean scheduleInline, boolean skipLocking) throws HoodieIOException {
-    if (scheduleInline) {
-      scheduleTableServiceInternal(cleanInstantTime, Option.empty(), TableServiceType.CLEAN);
-    }
     LOG.info("Cleaner started");
     final Timer.Context timerContext = metrics.getCleanCtx();
     LOG.info("Cleaned failed attempts if any");
     CleanerUtils.rollbackFailedWrites(config.getFailedWritesCleanPolicy(),
         HoodieTimeline.CLEAN_ACTION, () -> rollbackFailedWrites(skipLocking));
-    HoodieCleanMetadata metadata = createTable(config, hadoopConf).clean(context, cleanInstantTime, skipLocking);
+
+    // If there are pending clean operations, attempt them before scheduling the next clean. This prevents the
+    // next clean from deciding to clean the same files which might be under clean from pending operations.
+    HoodieCleanMetadata metadata = null;
+    HoodieTable table = createTable(config, hadoopConf);
```

Comment:
Shouldn't this be ensured by the cleaning caller ? Don't think we should put the logic here, and why only clean the first instant ?

Reply:
The caller does not know if the previous clean failed or succeed. For the caller to know this, it will have to check the timeline which is internal to HUDI and IMO should not be exposed to the caller.

Another way to reason about this is - "What is the HUDI Write Client API contract/guarantee?". When the caller calls clean() what does it guarantee?
1. That the clean will be scheduled, or
2. That the clean will eventually be completed, or
3. The clean may or may not complete

I might have misunderstood your point though.

>> why only clean the first instant ?
The logic (as I want to implement) cleans all leftover cleans and then performs an additional clean.

Reply:
> The caller does not know if the previous clean failed or succeed.

1. The caller could have know that if the cleaning task is executed in sequence
2. The caller does not need to know that if the cleaning plan is mutually exclusive in the fs view.

Reply:
yeah, I also feel the planner should ensure that clean plans are mutually exclusive. If there is a pending cleaning, it will eventually succeed. So, may be we can assume (for the most cases) once a clean plan is generated, it will get completed. Even on failure, it will get re-attempted right.

Reply:
>> yeah, I also feel the planner should ensure that clean plans are mutually exclusive.

This is what I have implemented by performing unfinished clean operations before attempting to generate a new clean plan. 

The other way of ensuring mutually exclusive clean plans by reading unfinished older clean plans which generating a new clean plan is overkill IMO.

Reply:
has some discussions w/ Ethan on this.here is what we feel. 
with spurious deletes, this should not an issue. only when the config is set to false, we need to worry about this. 
So, what we can do is, 
whenever hudi is looking to scheduling a clean, if there was another clean inflight found, no scheduling will happen. So, only when there are no pending inflight clean, a new clean will be scheduled. 
we don't want to trigger execution during scheduling phase. and hence.
the other option is: every time we wanted to plan cleaning, we might have to read every other inflight clean plan which we feel is too much to ask for, for the problem we are solving. 
bcoz, this happens only when metadata is enabled, and timeline server disabled and spurious deletes are not ignored.

Reply:
Sounds good to me.